docs/AsyncWrap/README.md

```diff
@@ -1,15 +1,275 @@
+
+`TCPConnectWrap` gets its information by setting properties on the handle
+object, like `address` and `port`. Those properties are read by the C++ layer,
+but can also be inspected from the AsyncWrap hooks. When the handle is create
+using `new TCPConnectWrap()` the `init` hook is called.
+
+A `oncomplete` property is also set, this is the callback for when the
+connection is made or failed. Just before calling `oncomplete` the `before` hook
+is called, just after the `after` hook is called.
+
+The `TCP` handle works exactly the same way, except that the information
+is passed as arguments to a method `.connect` and the `onread` function
+is called multiply times, thus it behaves like an event. This also means that
+the `before` and `after` hooks are called multiply times.
+
+Thus one should expect the hooks be called in the following order:
+
+```javascript
+init // TCPConnectWrap
+init // TCP
+=== tick ===
+before // TCPConnectWrap
+after // TCPConnectWrap
+=== tick ===
+before // TCP
+after // TCP
+=== tick ===
+before // TCP
+after // TCP
+=== tick ===
+...
+```
+
+_tick_ indicates there is at least one tick (as in `process.nextTick()`) between
+the text above and the text below.
+
+## The API
+
+At the moment there don't exist a high level API for AsyncWrap. It is simply exposed through `process.binding`:
+
+```javascript
+const asyncWrap = process.binding('async_wrap');
+```
+
+_Be warned that this API is not an official API and can change at any time, even if it's just patch
+update._
+
+To assign the hooks call:
+
+```javascript
+asyncWrap.setupHooks(init, before, after);
+function init(provider, parent) { /* consumer code */ }
+function before() { /* consumer code */ }
+function after() { /* consumer code */ }
+```
+
+Note that calling `asyncWrap.setupHooks` again, will overwrite the previous
+hooks.
+
+#### Enable and disable
+
+Because there is a small performance penalty in just calling a noop function,
+AsyncWrap is not enabled by default. To enable AsyncWrap call:
+
+```javascript
+asyncWrap.enable();
+```
+
+Note that handle objects created before AsyncWrap is enabled will never
+be tracked, even after `.enable()` is called. Similarly when AsyncWrap is
+disabled, the handle objects there are already being tracked will continue
+to be tracked. Finally there are some cases where handle objects have a parent.
+For example a TCP server creates TCP sockets, thus the socket handles have the
+server handle as a parent. In this case AsyncWrap should be enabled before
+the server handle is created, for the TCP sockets to be tracked.
+
+Disabling AsyncWrap can be done with:
+
+```javascript
+asyncWrap.disable();
+```
+
+#### The hooks
+
+Currently there are 3 hooks: `init`, `before` and `after`. The function
+signatures are quite similar. The `this` variable refers to the handle object,
+and `init` hook has two extra arguments `provider` and `parent`.
+
+```javascript
+function init(provider, parent) { this = handle; }
```

Comment:
I think it's fine to leave the function bodies empty. The rest of the text in that section adequately explains that `this` is the handle.